### PR TITLE
Wip.email multi event

### PR DESCRIPTION
**Pull request by RubenGeo**

### CODE REVIEW

**Review by gulfaraz — CHANGES_REQUESTED** [review]
Changes introduced in this PR,
- split email html into smaller components using [ejs](https://www.npmjs.com/package/ejs)
- separated email template code, email sending code, email content gathering code, and event specific code (massive respect for this refactor)
- rename `startDate` to `issuedDate`
- change in definition for finished events [`getEventsSummaryTriggerFinishedMail`](https://github.com/rodekruis/IBF-system/blob/ff66f245f8caf6517469176bdbb5a47a7922fef7/services/API-service/src/api/event/event.service.ts#L85-L117) and [`getFinishedEvent`](https://github.com/rodekruis/IBF-system/blob/c879e0673e78859f5eab05510c0ab6940eb5de38/services/API-service/src/api/notification/notification.service.ts#L75-L91)

Suggestions,
- Could we use an email HTML framework like [mjml](https://github.com/mjmlio/mjml)?
- Could we use a date library like [date-fns](https://date-fns.org/)?

Note for future refactors: Let's keep the new features in a separate PR from the refactor. If there are features added or removed, it's impossible to track if we've lost some functionality.

Would it be possible to document the notifications logic in a flow chart? Feel free to commit the flow chart image in the notifications folder. So it can be used as documentation coupled with the code.

TLDR; It looks okay - I haven't executed any code in this review. See comments for the merge blocker.

**Comment by RubenGeo** [comment]
Thanks for the feedback @gulfaraz! This was really what I was hoping to get out of this review

**Comment by RubenGeo** [comment]
Thanks for the review! Great that you summarized the changes!

> Note for future refactors: Let's keep the new features in a separate PR from the refactor. If there are features added or removed, it's impossible to track if we've lost some functionality.

I get this, and agree that it is sub-optimal

> Would it be possible to document the notifications logic in a flow chart? Feel free to commit the flow chart image in the notifications folder. So it can be used as documentation coupled with the code.

Sound like a good idea. I am creating an extra task for this on the task board AB#28230. 

> Could we use an email HTML framework like [mjml](https://github.com/mjmlio/mjml)?

I also created a separate task for this AB#28231. That probably would have been a better approach from the start. However considering the limit time to complete this work I'd rather first finish the API tests before implementing this.

File: services/API-service/src/api/event/event.service.ts [hunk 1]
```diff
@@ -908,7 +963,8 @@ export class EventService {
       where: whereFilters,
     });
 
-    //Below threshold events can be removed from this table after closing
+    // Below threshold events can be removed from this table after closing
+    // Below threshold events are warnings an not triggered. I do not know why they are removed here
```
[the remark this answers is not in the document]
Comment (RubenGeo) [reply services/API-service/src/api/event/event.service.ts:967]: But why are warnings handled differently than triggered events.

File: services/API-service/src/api/notification/notification.service.ts [hunk 2]
```diff
@@ -23,94 +23,103 @@ export class NotificationService {
...
+            disasterType,
+          );
+        }
       }
     }
-    return false;
   }
 
-  private async getNotifiableActiveEvent(
+  private async isNotifiableActiveEvent(
```
[the remark this answers is not in the document]
Comment (RubenGeo) [reply services/API-service/src/api/notification/notification.service.ts:122]: Yes, I do not know why but did not dare change it

File: services/API-service/src/api/event/event.service.ts [hunk 3]
```diff
@@ -69,38 +71,67 @@ export class EventService {
...
+    // I spend quite some time on trying to figure out what is the right query to get the event finished summary for the trigger closed email
+    // I came up with the following query but I am not sure if it is correct and how to test it properly
+    const sevenDaysAgo = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000);
+    const sixDaysAgo = new Date(Date.now() - 6 * 24 * 60 * 60 * 1000);
+    const whereFilters = {
+      endDate: Between(sevenDaysAgo, sixDaysAgo),
+      adminArea: In(countryAdminAreaIds),
+      disasterType: disasterType,
+      closed: true,
+    };
```
[the remark this answers is not in the document]
Comment (RubenGeo) [reply services/API-service/src/api/event/event.service.ts]: > the logic here is different from [getFinishedEvent](https://github.com/rodekruis/IBF-system/blob/c879e0673e78859f5eab05510c0ab6940eb5de38/services/API-service/src/api/notification/notification.service.ts#L75-L91). I suggest to confirm this new behaviour with the team before merge. Bcoz we've discussed this I agree your interpretation, the only thing that strongly recommend is to keep the check binary instead of a range. i.e. use LessThan instead of Between.

I did not get to checking this with the rest of the team unfortunately

**Review by gulfaraz — APPROVED** [review]

File: services/API-service/src/api/notification/email/email-template.service.ts [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+    );
+    return this.formatEmail(replaceKeyValues);
+  }
+
+  public async createHtmlForTriggerFinishedEmail(
+    country: CountryEntity,
+    disasterType: DisasterType,
+    finishedEvents: EventSummaryCountry[],
+    disasterTypeLabel: string,
+    _date: Date, // I am not sure if this is needed and for what it was used before
```
[the remark this answers is not in the document]
Comment (gulfaraz) [reply services/API-service/src/api/notification/email/email-template.service.ts:42]: your guess is as good as mine. we'll have to ask the author.
